lib/texd/cache.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module Texd
+  # Cache is a simple LRU cache with a double linked list.
+  # If the cache is full the last element from the list will be removed.
+  class Cache
+    class LinkedList
+      attr_reader :head
+
+      def add(node)
+        if @head
+          node.next       = @head
+          node.prev       = @head.prev
+          @head.prev.next = node
+          @head.prev      = node
+        else
+          node.next = node
+          node.prev = node
+        end
+
+        @head = node
+      end
+
+      def remove(node)
+        if node.next == node
+          # remove last element, should never happen
+          @head = nil
+        else
+          node.prev.next = node.next
+          node.next.prev = node.prev
+          @head          = node.next if node == @head
+        end
+      end
+    end
+
+    class Node
+      attr_reader :key
+      attr_accessor :value, :prev, :next
+
+      def initialize(key)
+        @key = key
+      end
+    end
+
+    attr_reader :list, :hash, :count, :capacity
```

Comment:
Should `hash` be publicly accessible? AFAICT, its usefulness is limited to the test suite.

Reply:
No does not need to be publicly accessible, I removed it.